File: maze.py
"""Maze data model and file helpers for graph pathfinding.

The maze is stored as a graph-friendly grid. Each open cell is a node, and
valid moves to neighbouring cells are graph edges. The model also keeps the
wall-based helpers used by the original coursework API.
"""

from __future__ import annotations

from dataclasses import dataclass, field
import json
from pathlib import Path
import random
from typing import Any, Dict, Iterable, List, Optional, Set, Tuple


Position = Tuple[int, int]

DIRECTIONS: Tuple[str, ...] = ("N", "E", "S", "W")
DIRECTION_VECTORS: Dict[str, Position] = {
    "N": (0, 1),
    "E": (1, 0),
    "S": (0, -1),
    "W": (-1, 0),
}


@dataclass
class Maze:
    width: int
    height: int
    blocked: Set[Position] = field(default_factory=set)
    horizontal_walls: Set[Position] = field(default_factory=set)
    vertical_walls: Set[Position] = field(default_factory=set)
    weights: Dict[Position, int] = field(default_factory=dict)
    start: Position = (0, 0)
    goal: Optional[Position] = None

    def __post_init__(self) -> None:
        if self.width <= 0 or self.height <= 0:
            raise ValueError("Maze dimensions must be positive")
        self.goal = self.goal if self.goal is not None else (self.width - 1, self.height - 1)
        self._add_outer_walls()
        self._validate_cells()

    def _add_outer_walls(self) -> None:
        for x in range(self.width):
            self.horizontal_walls.add((x, 0))
            self.horizontal_walls.add((x, self.height))
        for y in range(self.height):
            self.vertical_walls.add((0, y))
            self.vertical_walls.add((self.width, y))

    def _validate_cells(self) -> None:
        for cell in self.blocked | set(self.weights):
            if not self.in_bounds(cell):
                raise ValueError(f"Cell outside maze bounds: {cell}")
        for cell, weight in self.weights.items():
            if weight < 1:
                raise ValueError(f"Movement cost must be >= 1: {cell} = {weight}")
        if not self.is_open(self.start):
            raise ValueError(f"Start cell is blocked or outside bounds: {self.start}")
        if self.goal is not None and not self.is_open(self.goal):
            raise ValueError(f"Goal cell is blocked or outside bounds: {self.goal}")

    @property
    def open_cell_count(self) -> int:
        return self.width * self.height - len(self.blocked)

    def in_bounds(self, position: Position) -> bool:
        x, y = position
        return 0 <= x < self.width and 0 <= y < self.height

    def is_open(self, position: Position) -> bool:
        return self.in_bounds(position) and position not in self.blocked

    def movement_cost(self, position: Position) -> int:
        return self.weights.get(position, 1)

    def has_wall_between(self, first: Position, second: Position) -> bool:
        if not self.is_open(first) or not self.is_open(second):
            return True

        x1, y1 = first
        x2, y2 = second
        dx, dy = x2 - x1, y2 - y1

        if abs(dx) + abs(dy) != 1:
            raise ValueError("Wall checks require adjacent cells")
        if dx == 1:
            return (x2, y1) in self.vertical_walls
        if dx == -1:
            return (x1, y1) in self.vertical_walls
        if dy == 1:
            return (x1, y2) in self.horizontal_walls
        return (x1, y1) in self.horizontal_walls

    def neighbours(self, position: Position) -> List[Position]:
        result: List[Position] = []
        for direction in DIRECTIONS:
            dx, dy = DIRECTION_VECTORS[direction]
            candidate = (position[0] + dx, position[1] + dy)
            if self.is_open(candidate) and not self.has_wall_between(position, candidate):
                result.append(candidate)
        return result

    def get_walls(self, x: int, y: int) -> Tuple[bool, bool, bool, bool]:
        position = (x, y)
        walls: List[bool] = []
        for direction in DIRECTIONS:
            dx, dy = DIRECTION_VECTORS[direction]
            candidate = (x + dx, y + dy)
            walls.append(not self.is_open(candidate) or self.has_wall_between(position, candidate))
        return walls[0], walls[1], walls[2], walls[3]

    def add_horizontal_wall(self, x: int, horizontal_line: int) -> None:
        if not (0 <= x < self.width and 0 <= horizontal_line <= self.height):
            raise ValueError("Horizontal wall outside maze bounds")
        self.horizontal_walls.add((x, horizontal_line))

    def add_vertical_wall(self, y: int, vertical_line: int) -> None:
        if not (0 <= vertical_line <= self.width and 0 <= y < self.height):
            raise ValueError("Vertical wall outside maze bounds")
        self.vertical_walls.add((vertical_line, y))

    def clear_cell(self, position: Position, remove_adjacent_walls: bool = False) -> None:
        """Open a cell and optionally carve through its internal wall borders."""
        if not self.in_bounds(position):
            return
        self.blocked.discard(position)
        self.weights.pop(position, None)
        if remove_adjacent_walls:
            self.remove_adjacent_walls(position)

    def remove_adjacent_walls(self, position: Position) -> None:
        """Remove internal walls around a cell without deleting the outer border."""
        if not self.in_bounds(position):
            return
        x, y = position
        if y > 0:
            self.horizontal_walls.discard((x, y))
        if y < self.height - 1:
            self.horizontal_walls.discard((x, y + 1))
        if x > 0:
            self.vertical_walls.discard((x, y))
        if x < self.width - 1:
            self.vertical_walls.discard((x + 1, y))

    def to_ascii(
        self,
        path: Optional[Iterable[Position]] = None,
        visited: Optional[Iterable[Position]] = None,
    ) -> str:
        path_cells = set(path or [])
        visited_cells = set(visited or [])
        lines: List[str] = []
        for y in range(self.height - 1, -1, -1):
            chars: List[str] = []
            for x in range(self.width):
                position = (x, y)
                if position == self.start:
                    chars.append("S")
                elif position == self.goal:
                    chars.append("G")
                elif position in self.blocked:
                    chars.append("#")
                elif position in path_cells:
                    chars.append("*")
                elif position in visited_cells:
                    chars.append("+")
                elif position in self.weights:
                    chars.append(str(self.weights[position]))
                else:
                    chars.append(".")
            lines.append("".join(chars))
        return "\n".join(lines)


def create_maze(width: int = 5, height: int = 5) -> Maze:
    return Maze(width=width, height=height)


def add_horizontal_wall(maze: Maze, x_coordinate: int, horizontal_line: int) -> Maze:
    maze.add_horizontal_wall(x_coordinate, horizontal_line)
    return maze


def add_vertical_wall(maze: Maze, y_coordinate: int, vertical_line: int) -> Maze:
    maze.add_vertical_wall(y_coordinate, vertical_line)
    return maze


def get_dimensions(maze: Maze) -> Tuple[int, int]:
    return maze.width, maze.height


def get_walls(maze: Maze, x_coordinate: int, y_coordinate: int) -> Tuple[bool, bool, bool, bool]:
    return maze.get_walls(x_coordinate, y_coordinate)


def parse_ascii_maze(text: str) -> Maze:
    lines = [line.rstrip("\n") for line in text.splitlines() if line.strip()]
    if not lines:
        raise ValueError("Maze file is empty")
    if any(len(line) != len(lines[0]) for line in lines):
        raise ValueError("Maze must be rectangular")

    has_border = _has_solid_hash_border(lines)
    rows = lines[1:-1] if has_border else lines
    if has_border:
        rows = [row[1:-1] for row in rows]

    height = len(rows)
    width = len(rows[0]) if rows else 0
    blocked: Set[Position] = set()
    weights: Dict[Position, int] = {}
    start: Optional[Position] = None
    goal: Optional[Position] = None

    for file_y, row in enumerate(rows):
        y = height - 1 - file_y
        for x, char in enumerate(row):
            position = (x, y)
            if char == "#":
                blocked.add(position)
            elif char == "S":
                start = position
            elif char == "G":
                goal = position
            elif char.isdigit():
                # Only 2-9 are weighted terrain; '0' and '1' cost the same as a
                # plain open cell, so storing them would mislabel them as weighted.
                weight = int(char)
                if weight >= 2:
                    weights[position] = weight
            elif char not in {".", " "}:
                raise ValueError(f"Unsupported maze character {char!r} at {(x, y)}")

    return Maze(
        width=width,
        height=height,
        blocked=blocked,
        weights=weights,
        start=start or (0, 0),
        goal=goal or (width - 1, height - 1),
    )


def load_maze(path: str | Path) -> Maze:
    try:
        text = Path(path).read_text(encoding="utf-8")
    except OSError as exc:
        raise IOError(f"Cannot read maze file: {path}") from exc
    stripped = text.lstrip()
    if stripped.startswith("{"):
        return maze_from_json(json.loads(stripped))
    return parse_ascii_maze(text)


def save_maze(maze: Maze, path: str | Path) -> None:
    Path(path).write_text(json.dumps(maze_to_json(maze), indent=2), encoding="utf-8")


def maze_to_json(maze: Maze) -> Dict[str, Any]:
    return {
        "format": "graph-maze-pathfinder",
        "width": maze.width,
        "height": maze.height,
        "blocked": _positions_to_lists(maze.blocked),
        "horizontal_walls": _positions_to_lists(maze.horizontal_walls),
        "vertical_walls": _positions_to_lists(maze.vertical_walls),
        "weights": [
            {"x": x, "y": y, "cost": cost}
            for (x, y), cost in sorted(maze.weights.items())
        ],
        "start": list(maze.start),
        "goal": list(maze.goal) if maze.goal is not None else None,
    }


def maze_from_json(data: Dict[str, Any]) -> Maze:
    if data.get("format") != "graph-maze-pathfinder":
        raise ValueError("Unsupported maze JSON format")
    return Maze(
        width=int(data["width"]),
        height=int(data["height"]),
        blocked=_lists_to_positions(data.get("blocked", [])),
        horizontal_walls=_lists_to_positions(data.get("horizontal_walls", [])),
        vertical_walls=_lists_to_positions(data.get("vertical_walls", [])),
        weights={
            (int(item["x"]), int(item["y"])): int(item["cost"])
            for item in data.get("weights", [])
        },
        start=tuple(data.get("start", (0, 0))),  # type: ignore[arg-type]
        goal=tuple(data["goal"]) if data.get("goal") is not None else None,
    )


def create_perfect_maze(width: int, height: int, seed: Optional[int] = None) -> Maze:
    """Create a connected maze by carving passages with randomized DFS."""
    rng = random.Random(seed)
    maze = create_maze(width, height)

    for x in range(width):
        for y in range(1, height):
            maze.add_horizontal_wall(x, y)
    for y in range(height):
        for x in range(1, width):
            maze.add_vertical_wall(y, x)

    start = (0, 0)
    visited = {start}
    stack = [start]

    while stack:
        current = stack[-1]
        candidates = []
        for direction, (dx, dy) in DIRECTION_VECTORS.items():
            nxt = (current[0] + dx, current[1] + dy)
            if maze.in_bounds(nxt) and nxt not in visited:
                candidates.append((direction, nxt))
        if not candidates:
            stack.pop()
            continue

        direction, nxt = rng.choice(candidates)
        _remove_wall_between(maze, current, nxt, direction)
        visited.add(nxt)
        stack.append(nxt)

    maze.start = start
    maze.goal = (width - 1, height - 1)
    return maze


def add_random_weights(
    maze: Maze,
    density: float = 0.2,
    minimum: int = 2,
    maximum: int = 9,
    seed: Optional[int] = None,
) -> Maze:
    rng = random.Random(seed)
    density = min(1.0, max(0.0, density))
    candidates = [
        (x, y)
        for y in range(maze.height)
        for x in range(maze.width)
        if (x, y) not in {maze.start, maze.goal} and maze.is_open((x, y))
    ]
    rng.shuffle(candidates)
    target_count = int(round(len(candidates) * density))
    if density > 0 and candidates:
        target_count = max(1, target_count)
    for position in candidates[:target_count]:
        maze.weights[position] = rng.randint(minimum, maximum)
    return maze


def _has_solid_hash_border(lines: List[str]) -> bool:
    if len(lines) < 3 or len(lines[0]) < 3:
        return False
    return (
        all(char == "#" for char in lines[0])
        and all(char == "#" for char in lines[-1])
        and all(line[0] == "#" and line[-1] == "#" for line in lines)
    )


def _remove_wall_between(maze: Maze, first: Position, second: Position, direction: str) -> None:
    x1, y1 = first
    x2, y2 = second
    if direction == "E":
        maze.vertical_walls.discard((x2, y1))
    elif direction == "W":
        maze.vertical_walls.discard((x1, y1))
    elif direction == "N":
        maze.horizontal_walls.discard((x1, y2))
    elif direction == "S":
        maze.horizontal_walls.discard((x1, y1))


def _positions_to_lists(positions: Iterable[Position]) -> List[List[int]]:
    return [[x, y] for x, y in sorted(positions)]


def _lists_to_positions(items: Iterable[Iterable[int]]) -> Set[Position]:
    return {(int(x), int(y)) for x, y in items}

File: web/js/maze.js
/* Maze data model + generation — a faithful JS port of maze.py.
 * Grid is graph-friendly: open cells are nodes, valid moves are edges,
 * weighted cells affect entry cost. Origin is bottom-left (y grows upward).
 */

const DIRECTIONS = ["N", "E", "S", "W"];
const DIRECTION_VECTORS = {
  N: [0, 1],
  E: [1, 0],
  S: [0, -1],
  W: [-1, 0],
};

const key = (x, y) => x + "," + y;
const fromKey = (k) => k.split(",").map(Number);

// Mirror Python's round() (banker's rounding, half-to-even) so the JS port
// produces the same weighted-cell count as maze.py. Math.round rounds halves
// up, which diverges from Python at exact .5 values (e.g. 14.5 -> 14 not 15).
function pyRound(n) {
  const floor = Math.floor(n);
  const frac = n - floor;
  if (frac < 0.5) return floor;
  if (frac > 0.5) return floor + 1;
  return floor % 2 === 0 ? floor : floor + 1;
}

class Maze {
  constructor(width, height, opts = {}) {
    if (width <= 0 || height <= 0) throw new Error("Maze dimensions must be positive");
    this.width = width;
    this.height = height;
    this.blocked = new Set(opts.blocked || []);        // Set of "x,y"
    this.horizontalWalls = new Set(opts.horizontalWalls || []);
    this.verticalWalls = new Set(opts.verticalWalls || []);
    this.weights = new Map(opts.weights || []);        // Map "x,y" -> cost
    this.start = opts.start || [0, 0];
    this.goal = opts.goal || [width - 1, height - 1];
    this._addOuterWalls();
    this._validateCells();
  }

  // Mirror maze.py's _validate_cells so a malformed file fails fast and
  // identically across both implementations, instead of silently producing a
  // broken maze that only errors later at solve time.
  _validateCells() {
    for (const c of [...this.blocked, ...this.weights.keys()]) {
      const [x, y] = fromKey(c);
      if (!this.inBounds([x, y])) throw new Error("Cell outside maze bounds: " + c);
    }
    for (const cost of this.weights.values()) {
      if (cost < 1) throw new Error("Movement cost must be >= 1: " + cost);
    }
    if (!this.isOpen(this.start)) throw new Error("Start cell is blocked or outside bounds: " + this.start);
    if (this.goal && !this.isOpen(this.goal)) throw new Error("Goal cell is blocked or outside bounds: " + this.goal);
  }

  _addOuterWalls() {
    for (let x = 0; x < this.width; x++) {
      this.horizontalWalls.add(key(x, 0));
      this.horizontalWalls.add(key(x, this.height));
    }
    for (let y = 0; y < this.height; y++) {
      this.verticalWalls.add(key(0, y));
      this.verticalWalls.add(key(this.width, y));
    }
  }

  get openCellCount() {
    return this.width * this.height - this.blocked.size;
  }

  inBounds([x, y]) {
    return x >= 0 && x < this.width && y >= 0 && y < this.height;
  }

  isOpen(pos) {
    return this.inBounds(pos) && !this.blocked.has(key(pos[0], pos[1]));
  }

  movementCost(pos) {
    return this.weights.get(key(pos[0], pos[1])) || 1;
  }

  hasWallBetween(first, second) {
    if (!this.isOpen(first) || !this.isOpen(second)) return true;
    const [x1, y1] = first;
    const [x2, y2] = second;
    const dx = x2 - x1, dy = y2 - y1;
    if (Math.abs(dx) + Math.abs(dy) !== 1) throw new Error("Wall checks require adjacent cells");
    if (dx === 1) return this.verticalWalls.has(key(x2, y1));
    if (dx === -1) return this.verticalWalls.has(key(x1, y1));
    if (dy === 1) return this.horizontalWalls.has(key(x1, y2));
    return this.horizontalWalls.has(key(x1, y1));
  }

  neighbours(pos) {
    const result = [];
    for (const d of DIRECTIONS) {
      const [dx, dy] = DIRECTION_VECTORS[d];
      const cand = [pos[0] + dx, pos[1] + dy];
      if (this.isOpen(cand) && !this.hasWallBetween(pos, cand)) result.push(cand);
    }
    return result;
  }

  getWalls(x, y) {
    const walls = [];
    for (const d of DIRECTIONS) {
      const [dx, dy] = DIRECTION_VECTORS[d];
      const cand = [x + dx, y + dy];
      walls.push(!this.isOpen(cand) || this.hasWallBetween([x, y], cand));
    }
    return walls; // [N, E, S, W]
  }

  addHorizontalWall(x, line) {
    this.horizontalWalls.add(key(x, line));
  }
  addVerticalWall(y, line) {
    this.verticalWalls.add(key(line, y));
  }

  clearCell(pos, removeAdjacentWalls = false) {
    if (!this.inBounds(pos)) return;
    this.blocked.delete(key(pos[0], pos[1]));
    this.weights.delete(key(pos[0], pos[1]));
    if (removeAdjacentWalls) this.removeAdjacentWalls(pos);
  }

  removeAdjacentWalls(pos) {
    if (!this.inBounds(pos)) return;
    const [x, y] = pos;
    if (y > 0) this.horizontalWalls.delete(key(x, y));
    if (y < this.height - 1) this.horizontalWalls.delete(key(x, y + 1));
    if (x > 0) this.verticalWalls.delete(key(x, y));
    if (x < this.width - 1) this.verticalWalls.delete(key(x + 1, y));
  }

  clone() {
    const m = new Maze(this.width, this.height);
    m.blocked = new Set(this.blocked);
    m.horizontalWalls = new Set(this.horizontalWalls);
    m.verticalWalls = new Set(this.verticalWalls);
    m.weights = new Map(this.weights);
    m.start = [...this.start];
    m.goal = this.goal ? [...this.goal] : null;
    return m;
  }

  toJSON() {
    return {
      format: "graph-maze-pathfinder",
      width: this.width,
      height: this.height,
      blocked: [...this.blocked].map(fromKey),
      horizontal_walls: [...this.horizontalWalls].map(fromKey),
      vertical_walls: [...this.verticalWalls].map(fromKey),
      weights: [...this.weights].map(([k, cost]) => {
        const [x, y] = fromKey(k);
        return { x, y, cost };
      }),
      start: [...this.start],
      goal: this.goal ? [...this.goal] : null,
    };
  }
}

/* ── Seeded RNG (mulberry32) so seeds are reproducible like random.Random ── */
function mulberry32(seed) {
  let a = seed >>> 0;
  return function () {
    a |= 0; a = (a + 0x6D2B79F5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function createPerfectMaze(width, height, seed) {
  const rng = mulberry32(seed == null ? (Math.random() * 1e9) | 0 : seed);
  const maze = new Maze(width, height);

  // Fill all internal walls, then carve.
  for (let x = 0; x < width; x++)
    for (let y = 1; y < height; y++) maze.addHorizontalWall(x, y);
  for (let y = 0; y < height; y++)
    for (let x = 1; x < width; x++) maze.addVerticalWall(y, x);

  const start = [0, 0];
  const visited = new Set([key(0, 0)]);
  const stack = [start];

  const removeWallBetween = (first, second, direction) => {
    const [x1, y1] = first, [x2, y2] = second;
    if (direction === "E") maze.verticalWalls.delete(key(x2, y1));
    else if (direction === "W") maze.verticalWalls.delete(key(x1, y1));
    else if (direction === "N") maze.horizontalWalls.delete(key(x1, y2));
    else if (direction === "S") maze.horizontalWalls.delete(key(x1, y1));
  };

  while (stack.length) {
    const current = stack[stack.length - 1];
    const candidates = [];
    for (const d of DIRECTIONS) {
      const [dx, dy] = DIRECTION_VECTORS[d];
      const nxt = [current[0] + dx, current[1] + dy];
      if (maze.inBounds(nxt) && !visited.has(key(nxt[0], nxt[1]))) candidates.push([d, nxt]);
    }
    if (!candidates.length) { stack.pop(); continue; }
    const [direction, nxt] = candidates[Math.floor(rng() * candidates.length)];
    removeWallBetween(current, nxt, direction);
    visited.add(key(nxt[0], nxt[1]));
    stack.push(nxt);
  }

  maze.start = [0, 0];
  maze.goal = [width - 1, height - 1];
  return maze;
}

function addRandomWeights(maze, { density = 0.2, minimum = 2, maximum = 9, seed } = {}) {
  const rng = mulberry32(seed == null ? (Math.random() * 1e9) | 0 : seed);
  density = Math.min(1, Math.max(0, density));
  const sk = key(maze.start[0], maze.start[1]);
  const gk = maze.goal ? key(maze.goal[0], maze.goal[1]) : null;
  const candidates = [];
  for (let y = 0; y < maze.height; y++)
    for (let x = 0; x < maze.width; x++) {
      const k = key(x, y);
      if (k !== sk && k !== gk && maze.isOpen([x, y])) candidates.push([x, y]);
    }
  // Fisher-Yates with seeded rng
  for (let i = candidates.length - 1; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1));
    [candidates[i], candidates[j]] = [candidates[j], candidates[i]];
  }
  let target = pyRound(candidates.length * density);
  if (density > 0 && candidates.length) target = Math.max(1, target);
  for (const [x, y] of candidates.slice(0, target)) {
    maze.weights.set(key(x, y), minimum + Math.floor(rng() * (maximum - minimum + 1)));
  }
  return maze;
}

/* ── ASCII .mz parsing (for the embedded samples) ── */
function parseAsciiMaze(text) {
  const lines = text.split("\n").map((l) => l.replace(/\n$/, "")).filter((l) => l.trim().length);
  if (!lines.length) throw new Error("Maze file is empty");
  const w0 = lines[0].length;
  if (lines.some((l) => l.length !== w0)) {
    // pad ragged rows with spaces (open) so sample files stay tolerant
    for (let i = 0; i < lines.length; i++) lines[i] = lines[i].padEnd(w0, " ");
  }

  const solidBorder =
    lines.length >= 3 && lines[0].length >= 3 &&
    [...lines[0]].every((c) => c === "#") &&
    [...lines[lines.length - 1]].every((c) => c === "#") &&
    lines.every((l) => l[0] === "#" && l[l.length - 1] === "#");

  let rows = solidBorder ? lines.slice(1, -1).map((r) => r.slice(1, -1)) : lines;
  const height = rows.length;
  const width = rows.length ? rows[0].length : 0;
  const blocked = [];
  const weights = [];
  let start = null, goal = null;

  rows.forEach((row, fileY) => {
    const y = height - 1 - fileY;
    [...row].forEach((char, x) => {
      const k = key(x, y);
      if (char === "#") blocked.push(k);
      else if (char === "S") start = [x, y];
      else if (char === "G") goal = [x, y];
      // Only 2-9 are weighted terrain; '0'/'1' cost the same as an open cell.
      else if (/[2-9]/.test(char)) weights.push([k, parseInt(char, 10)]);
    });
  });

  return new Maze(width, height, {
    blocked,
    weights,
    start: start || [0, 0],
    goal: goal || [width - 1, height - 1],
  });
}

function mazeFromJSON(data) {
  if (data.format !== "graph-maze-pathfinder") throw new Error("Unsupported maze JSON format");
  return new Maze(data.width, data.height, {
    blocked: (data.blocked || []).map(([x, y]) => key(x, y)),
    horizontalWalls: (data.horizontal_walls || []).map(([x, y]) => key(x, y)),
    verticalWalls: (data.vertical_walls || []).map(([x, y]) => key(x, y)),
    weights: (data.weights || []).map((it) => [key(it.x, it.y), it.cost]),
    start: data.start || [0, 0],
    goal: data.goal || null,
  });
}

/* ── Embedded sample mazes (from sample_mazes/) ── */
const SAMPLE_MAZES = {
  "small_maze1.mz": `#########
#S..#...#
#.#.#.#.#
#.#...#.#
#.#####.#
#......G#
#########`,
  "maze1.mz": `#####################
#.....#.#...#.....#.#
###.###.#.#####.###.#
#.....#...#.....#...#
#####.#.###.#.#.###.#
#...#.#.....#.#.....#
###.#.#.#####.#.#.###
#.......#.....#.#.#.#
#.###.#.#####.#####.#
#...#.#.....#.......#
#.#.#.#####.#.#######
#.#.#.#.....#.......#
###.###.#####.#####.#
#.#.#.#.#...#.#...#.#
#.#.#.#####.#####.#.#
#...#...#...#.....#.#
#.#####.###.#.#####.#
#...#.........#.#.#.#
###.#.#####.###.#.#.#
#.......#...#.......#
#####################`,
  "maze2.mz": `###############
#S....#......#
#.###.#.####.#
#...#.#.#..#.#
###.#.#.#.##.#
#.#...#.#....#
#.#####.####.#
#.....#....#G#
#.###.####.#.#
#...#......#.#
###############`,
  "weighted_maze.mz": `###############
#S..2....#...G#
###.####.#.#..#
#...#..3...#..#
#.###.#####...#
#.....#...4...#
###############`,
};

function loadSampleMaze(name) {
  return parseAsciiMaze(SAMPLE_MAZES[name]);
}

window.MazeLib = {
  Maze, DIRECTIONS, DIRECTION_VECTORS, key, fromKey,
  createPerfectMaze, addRandomWeights, parseAsciiMaze, mazeFromJSON,
  SAMPLE_MAZES, loadSampleMaze,
};
